## Natural Language Processing - Summer Term 2024
### Hochschule Karlsruhe
### Lecturer: Prof. Dr. Jannik Strötgen
### Tutor: Paul Löhr

### Students:
### Yannick Walla (73005),
### Andreas Muth (66357),
### Vitali Reinhardt (72578),
### Moritz Wolf (91057)


# Exercise 05

- NLP Recap
- Entity Recognition

---

## Task 1 - NLP Recap:

## Q & A

Give answers to the following question. If you like, you can treat it as exam preparation, e.g., first try to solve the question without help of the slides ;) But you are obviously allowed to use the slides at any time


(1) Name three reasons why natural language processing (NLP) is challenging

**Ambiguities**:
- **Syntactic Ambiguity**: The exact same sentence can have multiple meanings
- **Anaphora Resolution**: It can sometimes be impossible to cleary identify which noun a pronoun refers to
- **Interpretation & World Knowledge**: Often understanding a sentence relies on unspoken context and common knowledge.

=> Having access to & understanding the right context helps with many ambiguities



(VL 1a p.30-34)

---

(2) What is tokenization?

- **Token**: The occurence of a word in a text
- **Tokenization**: Segmentation of an input stream into an ordered sequence of tokens
- **Tokenizer**: A system that splits texts into word tokens -> does tokenization

- Example sentence: "John likes Mary and Mary likes John."
    - Tokens: {"John", "likes", "Mary", "and", "Mary", "likes", "John", "."}

(VL 2 p.58)

---

(3) What is a word stem? Give the stem of the word "undoes"

- Stem: The common part to all inflected words
    - Does not have to be a proper word from the dictionary

    - => basically cut off at the maximum size that still is identical for word variants

- Example: undoes
    - Stem: undo (undone/undoing/undoable/undoer)

(VL 2 p.46)

---

(4) What is a word root? Give the root of the word "undoes"

- Root: The component of the word that
    - is common to a set of derived or inflected forms with all affixes removed
    - cannot be split further into meaningful components
    - carries the principal portion of meaning of the words
    - still not has to be found in the dictionary (-> Lemmatization)

- Example: undoes
    - Root: do (doing/done/redo/undo)

(VL 2 p.46)

---

(5) Why should we typically extract word stems/lemmas before preceding with text analysis?

- **Reduces Complexity**: Collapsing words into their stems/lemmas simplifies the dataset

- **Improves Consistency**: Different forms of a word are treated as a single item, improving the consistency of the analysis

- **Saves Resources**: Decreasing the number of unique words the system needs to process

- **Enhances Matching**: Improving search queries by being able to match different variants of the same word

- **Better Insights**: More accurate statistics from the text data if all variants of a word are aggregated

---

(6) What are stop-words? Why it makes sense to remove stop-words before preceding with text analysis?

Stop words are common words, like "the" and "and",  and are  removed before text analysis because they don't add much meaning. Removing them reduces noise in the data by focusing on more meaningful words, it improves efficiency by reducing computational complexity, it enhances the relevance of analyses like classification.

---

(7) If you had access to frequency statistics for a language, how could you create a list of stop words?

Analyze a large corpus of text data to identify frequently occurring words and set a threshold frequency below which words are considered stop words. Than  filter out any words that might be relevant in certain contexts

---

(8) Name a use-case in which we should NOT remove stop-words prior to text analysis.

Sentiment analysis

“I told you that she was not happy”

The result would be [‘told’, ‘happy’].

Meaning would be positive but it is not

---

(9) What is part-of-speech (POS) tagging? Why is it useful?

Part-of-speech (POS) tagging assigns grammatical categories to words in a sentence. POS tagging also disambiguates word meanings in context, improving accuracy in tasks like information retrieval.

---

(10) Explain how n-gram POS tagging works. What is the limitation of this method?

N-gram POS tagging combines n-grams with part-of-speech tagging to assign grammatical categories to words based on their context. It uses a statistical algorithm trained on labeled text data to predict POS tags by considering sequences of words. This approach captures more contextual information, enhancing accuracy, especially for words dependent on surrounding context. However, it faces limitations such as data sparsity due to the exponential increase in unique n-grams, limited context, and challenges in disambiguating words with multiple possible POS tags.

---

(11) What is parsing? Which two main types of parsing exist?

In NLP, parsing is one of the processes by which computers understand human language. Parsing involves breaking down a sentence into smaller pieces and identifying its grammatical structure.  So parsing is the process of analyzing the syntax and deeper structure to get the meaning.

There are two main techniques for parsing. These are top-down and bottom-up parsing.
In top-down parsing the parser tries to build the parse tree from the root node S down to the leaves. The main problem with this approach is that it explores options that do not lead to a complete parse.

The bottom-up approach starts with the words as input and then tries to build a tree from the words up, constantly applying grammatical rules one by one. It goes upwards towards the root. The problem with this method is backtracking. Where it explores a path that does not match the sentence.

---

(12) What is the difference between a context-free and a context-sensitive language? What is the difference between a context-free and a regular language?

Type-3 (Regular) are the most restricted, suitable only for simple patterns and sequences.

Type-2 (Context-free languages) offer more flexibility without context dependency, suitable for many programming and natural languages.

Type-1 (Context-sensitive languages) are the most expressive also capable of describing languages witha  complex context-dependent features.

Type 3
A → aB or 𝐴 → 𝑎
Where: 𝐴,𝐵 are non-terminals and 𝑎 is a terminal. It can contain at most one non-terminal, and if present, it must immediately follow a terminal.

Type 2
A→γ,
where 𝐴 is a non-terminal, and 𝛾 is any sequence of terminals and/or non-terminals (including the empty sequence). They don’t consider the context where a non-terminal is replaced.

Type 1
αAβ→αγβ, where: 𝐴 is a non-terminal and 𝛼,𝛽,𝛾 any sequence of symbols, including the empty sequence (except that 𝛾 must not be shorter than 𝐴). This ensures that the length of the string does not decrease by doing the application of a rule.
Type 1 are more powerful than context free by expressing the languages with intricate dependencies.


---

(13) Give an equivalent grammar in Chomsky Normal form for the Grammar G=(N,T,P,S) with N={S,A,B,C}, T={a,b,c}, P={S->ABC, A->a, B->b, C->c}


The Task:
N = {S, A, B, C} (non-terminal symbols)

T = {a, b, c} (terminal symbols)

P = {S -> ABC, A -> a, B -> b, C -> c} (production rules)

S is the start symbol


We need to convert the Formal Grammars to Chomsky Normal Form. This means to follow the rules from CNF.



ABC ist not Valid and we need to change it to:
S -> AX
X -> BC



We do not need to change this part: (A -> a, B -> b, C -> c)

N = {S, A, B, C, X} (non-terminal symbols)

T = {a, b, c} (terminal symbols)

P = {S -> AX, A -> a, B -> b, C -> c, X -> BC} (production rules)

S is the start symbol

---

(14) Explain how shift-reduce parsing works.

The Shift reduce is a bottom up parsing. It works by using a pushdown automaton understand a stack to get the CFL.

Do the following two operations

SHIFT = Put a word from the Input String to the stack

Reduce = Check if the words of the stack match a production rule on the right if yes replace it with the left-hand side.

Stop at S

---

(15) What is the grammar ambiguity problem?

Grammar ambiguity

The grammar ambiguity problem is the situation where a sentence or string can be parsed in multiple valid ways according to the same grammar, resulting in more than one valid parse or syntactic structure.

---

## Task 2 - Named Entity Recognition:

### Part 1: Automated Annotations

Use Spacy to annotate entities in the debates dataset (available as part of the JSON in the data directory). [Depending on your computer, the extraction may take some time. Therefore, you are allowed to restrict the size of the text, e.g. to the first 250 sentences. Feel free to use a larger share of data to get better insights.]

Tip: Use the en_core_web_sm corpus

Display the results in readable form, e.g. show the tagged entities for a reasonably sized part of the data, idealy alongside the original text.

In [2]:
# Tip for Spacy: In order to load a dataset, you might need to download the dataset via command line.
#Inside of a notebook, you can run commands with a !-mark

#Similar to: !python ...
!python -m spacy download en_core_web_sm

     ---------------------------------------- 0.0/12.8 MB ? eta -:--:--
      --------------------------------------- 0.2/12.8 MB 4.6 MB/s eta 0:00:03
     - -------------------------------------- 0.5/12.8 MB 6.8 MB/s eta 0:00:02
     -- ------------------------------------- 0.9/12.8 MB 6.3 MB/s eta 0:00:02
     --- ------------------------------------ 1.2/12.8 MB 7.0 MB/s eta 0:00:02
     ---- ----------------------------------- 1.6/12.8 MB 6.8 MB/s eta 0:00:02
     ------ --------------------------------- 1.9/12.8 MB 7.2 MB/s eta 0:00:02
     ------- -------------------------------- 2.3/12.8 MB 7.2 MB/s eta 0:00:02
     -------- ------------------------------- 2.6/12.8 MB 7.2 MB/s eta 0:00:02
     --------- ------------------------------ 2.9/12.8 MB 7.2 MB/s eta 0:00:02
     ---------- ----------------------------- 3.3/12.8 MB 7.3 MB/s eta 0:00:02
     ----------- ---------------------------- 3.6/12.8 MB 7.2 MB/s eta 0:00:02
     ------------ --------------------------- 4.0/12.8 MB 7

In [3]:
# read debates
import spacy
import json

with open('data/texts.json', 'r') as infile:
    data = json.load(infile)

In [4]:
# Your Code Submission goes here

content_debates = data['debates']

nlp = spacy.load("en_core_web_sm")
doc = nlp(content_debates)
first_250_sentences = list(doc.sents)[:250]

annotated_sentences = [{'text': sentence.text, 'entities': [(ent.text, ent.label_) for ent in sentence.ents]} for sentence in first_250_sentences]

for annotated_sentence in annotated_sentences:
    print("Original Text:", annotated_sentence['text'])
    print("Annotated Entities:", annotated_sentence['entities'])
    print()

Original Text:  Good evening from Hofstra University in Hempstead, New York.
Annotated Entities: [('evening', 'TIME'), ('Hofstra University', 'ORG'), ('Hempstead', 'GPE'), ('New York', 'GPE')]

Original Text: I am Lester Holt, anchor of "NBC Nightly News.”
Annotated Entities: [('Lester Holt', 'PERSON'), ('NBC Nightly News', 'ORG')]

Original Text: I want to welcome you to the first presidential debate.

Annotated Entities: [('first', 'ORDINAL')]

Original Text: The participants tonight are Donald Trump and Hillary Clinton.
Annotated Entities: [('tonight', 'TIME'), ('Donald Trump', 'PERSON'), ('Hillary Clinton', 'PERSON')]

Original Text: This debate is sponsored by the Commission on Presidential Debates, a nonpartisan, nonprofit organization.
Annotated Entities: [('the Commission on Presidential Debates', 'ORG')]

Original Text: The commission drafted tonight's format, and the rules have been agreed to by the campaigns.

Annotated Entities: [('tonight', 'TIME')]

Original Text: The 90-

### Part 2: Analysis

In real-word use cases, NER can be difficult. Analyze the following challenging examples and identify all cases in which the automated NER failed. For the tagging, you can use the method from part 1.

In [6]:
import json
with open('data/hard_data.json', 'r', encoding='utf-8') as infile:
    data = json.load(infile)

test_sentences = [s["sentence"] for s in data['test_sentences']]

Write the failure cases down here, then try to classify them into types of errors that the model makes. Discuss the results.
You can use a mixture of markdown and code if it suits your analysis.

In [7]:
import json
with open('data/hard_data.json', 'r', encoding='utf-8') as infile:
    data = json.load(infile)

test_sentences = [s["sentence"] for s in data['test_sentences']]

annotated_sentences = []
for sentence in test_sentences:
    doc = nlp(sentence)
    entities = [(ent.text, ent.label_) for ent in doc.ents]
    annotated_sentences.append({'text': sentence, 'entities': entities})

for annotated_sentence in annotated_sentences:
    print("Original Text:", annotated_sentence['text'])
    print("Annotated Entities:", annotated_sentence['entities'])
    print()

Original Text: Flight booking 1: “So, I would like to fly out sometime tonight and fly back in the evening in 4 days. From I’m looking to go to Denver. I’m flying out of San Francisco.”
Annotated Entities: [('1', 'CARDINAL'), ('tonight', 'TIME'), ('4 days', 'DATE'), ('Denver', 'GPE'), ('San Francisco', 'GPE')]

Original Text: Flight booking 2: “Okay, you got it so it looks like United Airlines leaves at 9:20 p.m. that is nonstop the flight duration is 2 hours and 28 minutes and is priced at $337.”
Annotated Entities: [('2', 'CARDINAL'), ('United Airlines', 'ORG'), ('9:20 p.m.', 'TIME'), ('2 hours and 28 minutes', 'TIME'), ('337', 'MONEY')]

Original Text: Flight booking 3: “I found a flight that leaves Seattle coming Monday at 7:35 a.m and arrives in Tampa at 4:10 p.m.”
Annotated Entities: [('3', 'CARDINAL'), ('Seattle', 'GPE'), ('coming Monday', 'DATE'), ('Tampa', 'GPE'), ('4:10 p.m.', 'TIME')]

Original Text: Hotel booking 1: “Park Hyatt Aviara resort Golf Club and Spa, it’s $279 per

# Entity Annotations Categories
(source: https://dataknowsall.com/blog/ner.html)

- PERSON:      People, including fictional.
- NORP:        Nationalities or religious or political groups.
- FAC:         Buildings, airports, highways, bridges, etc.
- ORG:         Companies, agencies, institutions, etc.
- GPE:         Countries, cities, states.
- LOC:         Non-GPE locations, mountain ranges, bodies of water.
- PRODUCT:     Objects, vehicles, foods, etc. (Not services.)
- EVENT:       Named hurricanes, battles, wars, sports events, etc.
- WORK_OF_ART: Titles of books, songs, etc.
- LAW:         Named documents made into laws.
- LANGUAGE:    Any named language.
- DATE:        Absolute or relative dates or periods.
- TIME:        Times smaller than a day.
- PERCENT:     Percentage, including ”%“.
- MONEY:       Monetary values, including unit.
- QUANTITY:    Measurements, as of weight or distance.
- ORDINAL:     “first”, “second”, etc.
- CARDINAL:    Numerals that do not fall under another type.

# Possible NER-Failure Classifications
- Incomplete Entity: When the recognized entity is not fully captured or is only partially identified
- Missing Entity: When an entity is present in the text but not detected by the NER system
- Misclassification of an Entity: When an entity is assigned to the wrong category
- Ambigious Entity: When the context of the entity is unclear, making it difficult to accurately classify
- False positive: When an entity is not present in the text, but is recognized and returned by the system

## Flight Booking 1:
    -
## Flight Booking 2:
    - Incomplete Entity: '337' as MONEY ==> should be: '$337'
    
## Flight Booking 3:
    - Missing Entity: '7:35 a.m' as TIME
    
## Hotel booking 1:
    - Incomplete Entity: 'Golf Club' is annotated as 'ORG' ==> should be: 'Park Hyatt Aviara resort Golf Club and Spa' as 'ORG'
    - Incomplete Entity: '279' as MONEY should be: '$279'
    
## Hotel booking 2:
    - Misclassification: 'Staybridge Suites Carlsbad' is annotated as 'WORK_OF_ART' ==> should be 'ORG'
    - Incomplete Entity: '145' as MONEY ==> should be '$145'
## Cinema 1:
    - Missing Entity: '1' as CARDINAL
    - Misclassification: 'The Mummy' is annotated as 'PERSON' ==> should be 'WORK_OF_ART'
    - Misclassification: 'Regal Davis Stadium 5' is annotated as 'FAC' ==> should be 'ORG'
## Cinema 2:
    - Missing Entity: 'Chips' as 'WORK_OF_ART'
    - Misclassification: 'Get Out' is annotated as 'EVENT' ==> should be 'WORK_OF_ART'
    - Misclassification: 'Snaps' is annotated as 'GPE' ==> should be 'WORK_OF_ART'
## Music 1:
    - Missing Entity: '1' as CARDINAL
    - Missing Entity: 'diamonds' as 'WORK_OF_ART'
    - Ambigous Entity Error: 'Rihanna Cover' is annotated as 'PERSON' ==> should be Cover by Rihanna (WORK_OF_ART) but could also be Rihanna (PERSON)
## Restaurant 1:
    - Missing Entity: '1' as CARDINAL
    - Missing Entity: 'IKEA' as 'ORG'
## Restaurant 2:
    - Missing Entity: '1' as CARDINAL
    - Incomplete Entity: 'Second Floor' is annotated as 'ORG' ==> should be 'Second Floor in Kitchen' as 'ORG'
    - Ambigous Entity Error: 'Second Floor in Kitchen' could be 'ORG' but could also be 'ORDINAL' (Second)
    - False-Positive Error: 'Kitchen' as 'GPE'
    - Ambigous Entity Error: 'New American' as 'NORP'
## Sports 1:
    - Missing Entity: '1–1' as CARDINAL


# Leads to 22 NER-Failures:
## 5x Incomplete Entity
## 8x Missing Entity
## 5x Misclassification
## 3x Ambigous Entity Error
## 1x False Positive Error

#### Submitting your results:

To submit your results, please:

- save this file, i.e., `ex??_assignment.ipynb`.
- if you reference any external files (e.g., images), please create a zip or rar archieve and put the notebook files and all referenced files in there.

**Remarks:**
    
- Do not copy any code from the Internet. In case you want to use publicly available code, please, add the reference to the respective code snippet.
- Check your code compiles and executes, even after you have restarted the Kernel.
- Submit your written solutions and the coding exercises within the provided spaces and not otherwise.
- Write the names of your partner and your name in the top section.